mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/MapboxTelemetry.java

```diff
@@ -0,0 +1,573 @@
+package com.mapbox.services.android.telemetry;
+
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.content.SharedPreferences;
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.content.pm.ServiceInfo;
+import android.location.Location;
+import android.location.LocationManager;
+import android.os.BatteryManager;
+import android.os.Build;
+import android.os.Handler;
+import android.support.annotation.NonNull;
+import android.support.v4.content.LocalBroadcastManager;
+import android.text.TextUtils;
+import android.util.DisplayMetrics;
+import android.view.WindowManager;
+
+import com.mapbox.services.android.telemetry.backoff.ExponentialBackoff;
+import com.mapbox.services.android.telemetry.connectivity.ConnectivityReceiver;
+import com.mapbox.services.android.telemetry.constants.TelemetryConstants;
+import com.mapbox.services.android.telemetry.http.TelemetryClient;
+import com.mapbox.services.android.telemetry.location.AndroidLocationEngine;
+import com.mapbox.services.android.telemetry.location.LocationEngine;
+import com.mapbox.services.android.telemetry.location.LocationEngineListener;
+import com.mapbox.services.android.telemetry.permissions.PermissionsManager;
+import com.mapbox.services.android.telemetry.service.TelemetryService;
+import com.mapbox.services.android.telemetry.utils.TelemetryUtils;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.Hashtable;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.UUID;
+import java.util.Vector;
+
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.Response;
+import okhttp3.internal.Util;
+import timber.log.Timber;
+
+/**
+ * This is the entry point to manage Mapbox telemetry
+ */
+public class MapboxTelemetry implements Callback, LocationEngineListener {
+
+  private static MapboxTelemetry instance;
+
+  private boolean initialized = false;
+  private Context context = null;
+  private String accessToken = null;
+  private String mapboxSessionId = null;
+  private long mapboxSessionIdLastSet = 0;
+  private String mapboxVendorId = null;
+  private DisplayMetrics displayMetrics = null;
+  private Intent batteryStatus = null;
+  private TelemetryClient client = null;
+  private Vector<Hashtable<String, Object>> events = new Vector<>();
+  private Timer timer = null;
+  private LocationEngine locationEngine = null;
+  private boolean withShutDown = false;
+  private Boolean telemetryEnabled = null;
+
+  /**
+   * Private constructor for configuring the single instance per app.
+   */
+  private MapboxTelemetry() {
+  }
+
+  /**
+   * Primary access method (using singleton pattern)
+   *
+   * @return MapboxTelemetry
+   */
+  public static MapboxTelemetry getInstance() {
```

Comment:
synchronised keyword for thread safety or annotate the method to be called only from main thread?